File: SecondYear/AvantikaSharma2307_AvantikaSharma_2226cseai1008_2/Week_2/parenthesis.py
def generate(n,open,close,current,result):
    if(len(current)==2*n):
        result.append(current)
        return
    if(open<n):
        generate(n,open+1,close,current +"(",result)
    if(close<open):
        generate(n,open,close+1,current+")",result)        




def validParenthesis(n: int) -> int:
    result=[]
    generate(n,0,0,"",result)
    return result

File: SecondYear/29deepanshutyagi_DeepanshuTyagi_2226cs1195_2/Week_2/parenthesis.cpp
#include <vector>
#include <string>

void generateParenthesis(int n, int open, int close, std::string current, std::vector<std::string> &result) {
    if (current.length() == 2 * n) {
        result.push_back(current);
        return;
    }

    if (open < n) {
        generateParenthesis(n, open + 1, close, current + '(', result);
    }

    if (close < open) {
        generateParenthesis(n, open, close + 1, current + ')', result);
    }
}

std::vector<std::string> validParenthesis(int n) {
    std::vector<std::string> result;
    generateParenthesis(n, 0, 0, "", result);
    return result;
}
